File: sdks/ruby/test/smoke_test.rb
# frozen_string_literal: true

# Offline smoke test — validates request shaping, HMAC signing and webhook
# verification without hitting the network. Run: ruby test/smoke_test.rb
require "openssl"
require "json"
require_relative "../lib/pushfy"

$passed = 0
def ok(name)
  puts "  ✓ #{name}"
  $passed += 1
end

def assert(cond, msg = "assertion failed")
  raise msg unless cond
end

# Records the last request and returns a canned response. Compatible with the
# Client's injectable transport: #call(url, method:, headers:, body:) -> Hash.
class MockHttp
  attr_reader :calls

  def initialize(status: 200, body: "")
    @status = status
    @body = body.is_a?(String) ? body : JSON.generate(body)
    @calls = []
  end

  def call(url, method:, headers:, body:)
    @calls << { url: url, method: method, headers: headers, body: body }
    { status: @status, body: @body }
  end
end

begin
  # 1. HMAC signing matches the documented recipe exactly.
  ts = 1_752_345_600
  body = '{"user_ext_id":"user-42"}'
  secret = "pas_test"
  sig = Pushfy::Hmac.sign(method: "post", path: "/v1/conversations", body: body, secret: secret, timestamp: ts)
  bh = OpenSSL::Digest::SHA256.hexdigest(body)
  base = "#{ts}\nPOST\n/v1/conversations\n#{bh}"
  expected = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha256"), secret, base)
  assert(sig[:signature] == expected, "HMAC signature mismatch")
  assert(sig[:timestamp] == ts.to_s, "timestamp mismatch")
  ok("HMAC signature matches canonical base string (ts\\nMETHOD\\nroute\\nsha256hex(body))")

  # 2. sms.send hits /webapi with Bearer auth and normalized digits.
  http = MockHttp.new(body: [{ "id" => "x", "phone" => "5511999999999", "ext_id" => "x" }])
  pushfy = Pushfy::Client.new(api_token: "YOUR_API_TOKEN", http: http)
  res = pushfy.sms.send(to: "+55 (11) 99999-9999", text: "Hi", ext_id: "x")
  call = http.calls[0]
  assert(call[:url].end_with?("/webapi"), "URL is /webapi")
  assert(call[:method] == "POST", "method POST")
  assert(call[:headers]["Authorization"] == "Bearer YOUR_API_TOKEN", "Bearer header")
  sent = JSON.parse(call[:body])
  assert(sent["messages"][0]["destinations"][0]["to"] == "5511999999999", "phone digits normalized")
  assert(sent["messages"][0]["text"] == "Hi", "text sent")
  assert(res[0]["ext_id"] == "x", "array response parsed")
  ok("sms.send shapes /webapi request and parses the array response")

  # 3. balance.get parses the formatted string "1.500" -> 1500.
  http = MockHttp.new(body: { "saldo" => "1.500" })
  pushfy = Pushfy::Client.new(api_token: "t", http: http)
  b = pushfy.balance.get
  assert(b == { raw: "1.500", balance: 1500 }, "balance parse: #{b.inspect}")
  ok('balance.get parses {"saldo":"1.500"} -> 1500')

  # 4. conversations.open signs with X-PA-* headers and routes via ?r=.
  http = MockHttp.new(body: { "ok" => true, "conversation_id" => 1, "status" => "bot" })
  pushfy = Pushfy::Client.new(pa_key: "pak_x", pa_secret: "pas_x", http: http)
  pushfy.conversations.open(user_ext_id: "user-42", name: "Ana")
  call = http.calls[0]
  assert(call[:url].include?("r=%2Fv1%2Fconversations"), "route sent via ?r=")
  assert(call[:headers]["X-PA-Key"] == "pak_x", "X-PA-Key header set")
  assert(call[:headers]["X-PA-Signature"] && call[:headers]["X-PA-Timestamp"], "signature + timestamp set")
  ok("conversations.open signs request with X-PA-* headers")

  # 5. push server call uses X-PUSH-* headers.
  http = MockHttp.new(body: { "ok" => true })
  pushfy = Pushfy::Client.new(push_key: "pushk_x", push_secret: "pss_x", http: http)
  pushfy.push.campaigns.send(88)
  call = http.calls[0]
  assert(call[:url].include?("r=%2Fv1%2Fpush%2Fcampaigns%2F88%2Fsend"), "push route encoded")
  assert(call[:headers]["X-PUSH-Key"] == "pushk_x", "X-PUSH-Key header set")
  ok("push.campaigns.send signs with X-PUSH-* headers")

  # 6. webhook verification: raw vs prefixed schemes.
  secret = "WEBHOOK_SECRET"
  payload = '{"eid":"evt_1","event":"handoff.requested"}'
  hex = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha256"), secret, payload)
  assert(Pushfy::Webhooks.conversations(payload: payload, signature: hex, secret: secret) == true,
         "raw (X-PA) accepts bare hex")
  assert(Pushfy::Webhooks.push(payload: payload, signature: "sha256=#{hex}", secret: secret) == true,
         "prefixed accepts sha256= hex")
  assert(Pushfy::Webhooks.push(payload: payload, signature: hex, secret: secret) == false,
         "push requires sha256= prefix")
  assert(Pushfy::Webhooks.conversations(payload: payload, signature: "deadbeef", secret: secret) == false,
         "bad signature rejected")
  assert(Pushfy::Webhooks.messaging(payload: payload, signature: nil, secret: secret) == false,
         "nil signature rejected")
  ok("webhook verify handles raw (X-PA) and prefixed (X-Push/X-Pushfy) schemes")

  # 7. errors: 401 -> AuthenticationError.
  http = MockHttp.new(status: 401, body: { "ok" => false, "error" => "unauthorized" })
  pushfy = Pushfy::Client.new(api_token: "bad", http: http)
  caught = nil
  begin
    pushfy.sms.send(to: "5511999999999", text: "x")
  rescue Pushfy::PushfyError => e
    caught = e
  end
  assert(caught.is_a?(Pushfy::AuthenticationError), "AuthenticationError class")
  assert(caught.status == 401, "status 401")
  assert(caught.code == "unauthorized", "code parsed")
  ok("401 response raises AuthenticationError")

  # 8. errors: 429 -> RateLimitError.
  http = MockHttp.new(status: 429, body: { "error" => "rate_limited" })
  pushfy = Pushfy::Client.new(api_token: "t", http: http)
  caught = nil
  begin
    pushfy.sms.send(to: "5511999999999", text: "x")
  rescue Pushfy::PushfyError => e
    caught = e
  end
  assert(caught.is_a?(Pushfy::RateLimitError), "RateLimitError class")
  ok("429 response raises RateLimitError")

  puts "\n#{$passed} checks passed."
rescue StandardError => e
  warn "\nSMOKE TEST FAILED: #{e.message}"
  warn e.backtrace.first(5).join("\n")
  exit 1
end

File: sdks/java/src/test/java/com/pushfy/SmokeTest.java
package com.pushfy;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Offline smoke test — validates HMAC signing, webhook verification, balance
 * parsing and request shaping without hitting the network.
 *
 * <p>Runs with a bare JDK (no test framework required):
 * <pre>
 *   javac -d out $(find src/main/java -name '*.java') src/test/java/com/pushfy/SmokeTest.java
 *   java  -cp out com.pushfy.SmokeTest
 * </pre>
 */
public final class SmokeTest {

    private static int passed = 0;

    public static void main(String[] args) {
        // 1. HMAC signing matches the documented recipe exactly:
        //    base = ts\nMETHOD\nroute\nsha256hex(body)
        {
            long ts = 1752345600L;
            String body = "{\"user_ext_id\":\"user-42\"}";
            String secret = "pas_test";
            Hmac.Signed signed = Hmac.sign("post", "/v1/conversations", body, secret, ts);
            String bh = Hmac.sha256Hex(body);
            String base = ts + "\nPOST\n/v1/conversations\n" + bh;
            String expected = Hmac.hmacSha256Hex(base, secret);
            assertEq(expected, signed.signature, "HMAC signature matches canonical base string");
            assertEq(String.valueOf(ts), signed.timestamp, "HMAC preserves the supplied timestamp");
        }

        // 2. Known SHA-256 vector for the empty string (guards the hex/digest path).
        {
            assertEq("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
                    Hmac.sha256Hex(""), "sha256(\"\") matches the known vector");
        }

        // 3. Webhook verification: raw (X-PA) vs prefixed (X-Push/X-Pushfy) schemes.
        {
            String secret = "WEBHOOK_SECRET";
            String payload = "{\"eid\":\"evt_1\",\"event\":\"handoff.requested\"}";
            String hex = Hmac.hmacSha256Hex(payload, secret);
            assertTrue(Webhooks.conversations(payload, hex, secret),
                    "conversations() accepts raw hex signature");
            assertTrue(Webhooks.push(payload, "sha256=" + hex, secret),
                    "push() accepts sha256= prefixed signature");
            assertTrue(Webhooks.messaging(payload, "sha256=" + hex, secret),
                    "messaging() accepts sha256= prefixed signature");
            assertFalse(Webhooks.push(payload, hex, secret),
                    "push() requires the sha256= prefix");
            assertFalse(Webhooks.conversations(payload, "sha256=" + hex, secret),
                    "conversations() rejects the sha256= prefix (raw scheme)");
            assertFalse(Webhooks.conversations(payload, "deadbeef", secret),
                    "conversations() rejects a bogus signature");
            assertFalse(Webhooks.messaging(payload, null, secret),
                    "verify() rejects a null signature");
        }

        // 4. balance.get parses the formatted string {"saldo":"1.500"} -> 1500.
        {
            Map<String, Object> res = new LinkedHashMap<>();
            res.put("saldo", "1.500");
            Pushfy.Balance b = Pushfy.balanceFrom(res);
            assertEq("1.500", b.raw, "balance keeps the raw string");
            assertTrue(b.balance != null && b.balance == 1500L, "balance parses \"1.500\" -> 1500");
        }

        // 5. Phone normalization strips non-digits (matches the Node SDK).
        {
            assertEq("5511999999999", Pushfy.digits("+55 (11) 99999-9999"),
                    "digits() normalizes a formatted phone number");
        }

        // 6. toMessage shapes the /webapi message body correctly.
        {
            Map<String, Object> m = Pushfy.toMessage("+55 (11) 99999-9999", "Hi", "x", null);
            Object dests = m.get("destinations");
            assertTrue(dests instanceof List, "destinations is a list");
            @SuppressWarnings("unchecked")
            Map<String, Object> d0 = (Map<String, Object>) ((List<Object>) dests).get(0);
            assertEq("5511999999999", String.valueOf(d0.get("to")), "message phone normalized");
            assertEq("Hi", String.valueOf(m.get("text")), "message text preserved");
            assertEq("x", String.valueOf(m.get("ext_id")), "message ext_id preserved");
        }

        // 7. JSON round-trips the request/response shapes we rely on.
        {
            Map<String, Object> obj = new LinkedHashMap<>();
            obj.put("s", "a\"b\\c\n");
            obj.put("n", 1500L);
            obj.put("b", Boolean.TRUE);
            obj.put("nil", null);
            List<Object> arr = new ArrayList<>();
            arr.add("x");
            arr.add(42L);
            obj.put("arr", arr);
            String enc = Json.encode(obj);
            Object back = Json.parse(enc);
            assertTrue(back instanceof Map, "parsed JSON is an object");
            @SuppressWarnings("unchecked")
            Map<String, Object> pm = (Map<String, Object>) back;
            assertEq("a\"b\\c\n", String.valueOf(pm.get("s")), "JSON string escapes round-trip");
            assertTrue(pm.get("n") instanceof Long && (Long) pm.get("n") == 1500L,
                    "JSON integer round-trips as Long");
            assertTrue(Boolean.TRUE.equals(pm.get("b")), "JSON boolean round-trips");
            assertTrue(pm.containsKey("nil") && pm.get("nil") == null, "JSON null round-trips");
            assertEq("[\"x\",42]", Json.encode(pm.get("arr")), "JSON array round-trips");
        }

        // 8. Array response parsing (e.g. sms.send result).
        {
            Object parsed = Json.parse("[{\"id\":\"x\",\"ext_id\":\"x\"}]");
            assertTrue(parsed instanceof List, "top-level array parses to a List");
        }

        System.out.println();
        System.out.println(passed + " checks passed.");
    }

    private static void assertEq(String expected, String actual, String name) {
        if (expected == null ? actual == null : expected.equals(actual)) {
            ok(name);
        } else {
            fail(name + " (expected=" + expected + " actual=" + actual + ")");
        }
    }

    private static void assertTrue(boolean cond, String name) {
        if (cond) {
            ok(name);
        } else {
            fail(name);
        }
    }

    private static void assertFalse(boolean cond, String name) {
        assertTrue(!cond, name);
    }

    private static void ok(String name) {
        System.out.println("  [ok] " + name);
        passed++;
    }

    private static void fail(String name) {
        System.out.println("  [FAIL] " + name);
        throw new AssertionError("SMOKE TEST FAILED: " + name);
    }
}
